components/places/src/db/db.rs

```diff
@@ -9,22 +9,123 @@
+    encryption_key: Option<String>,
+    write_connection: Mutex<Cell<Option<PlacesDb>>>,
+    // XXX - TODO - the "sync" connection.
+}
+
+impl PlacesAPI {
+    /// Create a new PlacesAPI backed by a file on disk.
+    pub fn new(db_name: impl AsRef<Path>, encryption_key: Option<&str>) -> Result<Self> {
+        // We always create a new read-write connection for an initial open so
+        // we can create the schema and/or do version upgrades.
+        // XXX - pathbuf usage here is wrong.
+        let connection = PlacesDb::open(
+            &db_name.as_ref().to_path_buf(),
+            encryption_key.as_ref().map(|x| &**x),
+            ConnectionType::ReadWrite,
+        )?;
+        Ok(PlacesAPI {
+            db_name: db_name.as_ref().to_path_buf(),
+            encryption_key: encryption_key.map(|x| x.to_string()),
+            write_connection: Mutex::new(Cell::new(Some(connection))),
+        })
+    }
+
+    /// Create a new memory-based PlacesAPI. You must provide a name, but
+    /// you are still able to have a single writer and many reader connections
+    /// to the same memory DB open.
+    pub fn new_memory(db_name: &str, encryption_key: Option<&str>) -> Result<Self> {
+        let name = PathBuf::from(format!("file:{}?mode=memory&cache=shared", db_name));
+        Self::new(&name, encryption_key)
+    }
+
+    /// Open a connection to the database.
+    pub fn open_connection(&self, conn_type: ConnectionType) -> Result<PlacesDb> {
+        let ec = self.encryption_key.as_ref().map(|x| &**x); // black magic!
+        match conn_type {
+            ConnectionType::ReadOnly => {
+                // make a new one - we can have as many of these as we want.
+                PlacesDb::open(self.db_name.clone(), ec, ConnectionType::ReadOnly)
+            }
+            ConnectionType::ReadWrite => {
+                // We only allow one of these.
+                let cell = self.write_connection.lock().unwrap();
+                match cell.replace(None) {
+                    None => Err(ErrorKind::ConnectionAlreadyOpen.into()),
+                    Some(db) => Ok(db),
+                }
+            }
+            ConnectionType::Sync => {
+                // ideally we'd enforce this in the same way as write_connection
+                PlacesDb::open(self.db_name.clone(), ec, ConnectionType::Sync)
+            }
+        }
+    }
+
+    /// Close a connection to the database. If the connection is the write
+    /// connection, you can re-fetch it using open_connection.
+    pub fn close_connection(&self, connection: PlacesDb) -> Result<()> {
+        if connection.conn_type == ConnectionType::ReadWrite {
+            // We only allow one of these.
+            let cell = self.write_connection.lock().unwrap();
+            assert!(cell.replace(Some(connection)).is_none());
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
 pub struct PlacesDb {
     pub db: Connection,
+    pub conn_type: ConnectionType,
```

Comment:
This should probably not be `pub`, since someone changing would break things. Admittedly, we tend to pass &PlacesDb around in practice, but still.